dopamine/discrete_domains/run_experiment.py

```diff
@@ -600,18 +609,57 @@ def _run_iterations(self):
     time an iteration completes a new one starts until the right number of
     iterations is run.
     """
-    # TODO(aarg): Change the thread management to an implementation with queues.
-    threads = []
+    # TODO(aarg): Change the thread management to an implementation with queues
+    # and with a fix number of thread workers. With the revamp, training might
+    # go inline to this method.
+    training_worker = threading.Thread(target=self._run_training_steps)
+    training_worker.start()
+    threads = [training_worker]
     for iteration in range(self._start_iteration, self._num_iterations):
-      self._running_iterations.acquire()
+      self._running_iterations.put(1)
       thread = threading.Thread(
           target=self._run_one_iteration, args=(iteration,))
       thread.start()
       threads.append(thread)
-    # Wait for all running iterations to complete.
+
+    # Wait for all tasks to complete.
+    self._running_iterations.join()
+    self._remaining_training_steps.join()
+    # Indicate training step thread to stop.
+    self._remaining_training_steps.put(None)
+    # Wait for all running threads to complete.
     for thread in threads:
       thread.join()
 
+  def _begin_episode(self, observation):
+    # Increments training steps and blocks if training is too slow.
+    self._training_step()
+    return self._agent.begin_episode(observation, training=False)
+
+  def _step(self, reward, observation):
+    # Increments training steps and blocks if training is too slow.
+    self._training_step()
+    return self._agent.step(reward, observation, training=False)
+
+  def _training_step(self):
```

Comment:
nit: `_training_step` isn't very descriptive for this method's functionality.  It adds a training step to the training_steps queue.  Maybe an alternative is `_enqueue_training_step`?

Reply:
done.